doc/error-handling.md

```diff
@@ -0,0 +1,324 @@
+
+
+# Error Handling
+Errors are considered to be system states that should not be reached regularly and usually are the result of an external failure, such as when the OS is unable to provide a certain resource (e.g. a semaphore) or an application does not respond. In contrast, regular behaviour such as a receiver receiving no data when none was sent is not an error. On the other hand, losing data that was sent would be considered an error.
+
+There are two general approaches to deal with errors:
+1. using exceptions 
+
+2. return codes combined with control flow statements and a central instance to handle errors that cannot be mitigated otherwise (the error handler). 
+
+In Iceoryx we use the latter approach.
+
+## Exceptions
+The use of exceptions in Iceoryx is prohibited, including third party code that may throw them. This is due to the following reasons:
+* In many implementations exception handling requires dynamic memory (even when the exceptions themselves are not generated dynamically via e.g. new).
+* Exception handling is not deterministic with respect to runtime.
+* Exception handling may cause a (slight) runtime overhead even when no exceptions are thrown.
+* In general it is not possible to incorporate (complete) information about all exceptions that can be thrown in the signature of functions. This makes it hard for the caller to decide whether something some exception needs to be handled.
+* Overuse of exceptions often leads to convoluted try-catch blocks everywhere which makes the code hard to maintain.
+
+### Use of noexcept
+All functions are marked as *noexcept*. Note that this does not mean that exceptions cannot be thrown inside such a function. Instead when an exeption is thrown inside the function (either directly or indirectly by another function) this exception is not propagated further and *std::terminate* will be invoked, calling the terminate handler. The goal is that this cannot happen under any circumstances during runtime.
+
+To this end it is necessary to eliminate the use of all potentially throwing (third party) functions throughout the codebase. Since many STL functions may throw, these cannot be used either and their functionality needs to be reimplemented without the use of exceptions. In particular anything allocating dynamic memory may throw a *std::bad_alloc* exception when memory is exhausted.
+
+### Alternatives to exceptions
+As an alternative to exceptions we use the error handler and a variation of return codes in the form of **cxx::expected**, described below. cxx::expected can be used to communicate the error to the caller, who has to decide whether to handle the error itself or propagate it further (e.g. as another cxx::expected). Error handling itself is performed by the error handler which handles errors occuring in the subcomponents of iceoryx::posh.
```

Comment:
I'm kind of a fan of using code style for all code related keywords, e.g. `cxx::expected`, `LogErr`, `std::bad_alloc`, etc. Makes a doc easier to read/understand and is more distinguishable than _italic_ or **bold**.. Especially as it's not consistent throughout this document.

Reply:
Yes, will go over this and change it. Bold is too much in my opinion though and italics somehow a little weak.